### PR TITLE
Support `no_std` targets

### PR DESCRIPTION
**Pull request by mkroening**
### Summary of the PR

This adds `no_std` support by
- adding a default `std` feature for disabling `impl Error` on `no_std`.
- adding the Hashbrown dependency for `no_std` hash sets

Hashbrown also provides the `HashSet` implementation in `std`. `std`'s `HashSet` is not in liballoc yet, though.
By default, Hashbrown uses AHash as the default hasher. AHash is much faster than `std`'s SipHash, but is not as HashDoS resistant. That seemed not relevant to me, though.

Closes https://github.com/rust-vmm/vm-fdt/issues/61

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [ ] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

### CODE REVIEW

**Comment by mkroening** [comment]
I don't know how to modify CI to test this. Locally, I ran the following for testing:

```
cargo test
cargo test --no-default-features
cargo check --target x86_64-unknown-none --no-default-features
```

File: Cargo.toml
Comment (andreeaflorescu) [inline Cargo.toml:12]: We are trying to keep the dependencies to a minimum, and for this crate we have none. Is there a way to have this at least as an optional dependency?

File: src/writer.rs
Comment (andreeaflorescu) [inline src/writer.rs:82]: Maybe we can implement https://doc.rust-lang.org/beta/core/error/trait.Error.html for no_std? WDYT?

**Comment by andreeaflorescu** [comment]
> I don't know how to modify CI to test this. Locally, I ran the following for testing:
> 
> ```
> cargo test
> cargo test --no-default-features
> cargo check --target x86_64-unknown-none --no-default-features
> ```

To test this as part of the CI you'll need to add a custom pipeline. Here is an example of a custom pipeline in another repository: https://github.com/rust-vmm/vm-virtio/tree/main/.buildkite. Once we get close with merging this PR, I can make the pipeline run on all PRs by configuring it in Buildkite.

**Review by andreeaflorescu — COMMENTED** [review]
LGTM, I have just some questions. Thanks for handling this!

File: Cargo.toml
Comment (mkroening) [reply Cargo.toml:12]: There is no _nice_ way of doing that since Cargo features are additive and `feature = "std"` would only add things.

We could add an `alloc` feature that has to be enabled explicitly. That way `cargo check --no-default-features` without any features will fail though, and one has to enable `alloc` explicitly on `no_std`.

File: src/writer.rs
Comment (mkroening) [reply src/writer.rs:82]: I'd like to avoid requiring `nightly` for `no_std`. We could add another `nightly` feature, but I don't see a big benefit in this.

File: Cargo.toml
Comment (danielverkamp) [reply Cargo.toml:12]: I guess the other alternative would be to just skip duplicate phandle checking in the `no_std` version so we would not need a `HashSet` at all; that would make it the caller's responsibility to ensure phandles are unique (like it was before we added this check). Maybe tying that to `no_std` would be too confusing, though.

Comment (mkroening) [reply Cargo.toml:12]: I think, I'd rather not have this compile on `no_std` without a feature than removing functionality from this crate. I'll implement this shortly.

Comment (mkroening) [reply Cargo.toml]: Are you okay with the solution I implemented? :)

**Comment by mkroening** [comment]
> To test this as part of the CI you'll need to add a custom pipeline. Here is an example of a custom pipeline in another repository: https://github.com/rust-vmm/vm-virtio/tree/main/.buildkite. Once we get close with merging this PR, I can make the pipeline run on all PRs by configuring it in Buildkite.

I added a CI config. 👍

File: Cargo.toml
Comment (andreeaflorescu) [reply Cargo.toml]: I think it's fine as it is, I was just wondering if it's possible.

File: src/writer.rs
Comment (andreeaflorescu) [reply src/writer.rs:85]: Oooops, I didn't see that you need nightly for that. That's a good call!

**Review by andreeaflorescu — APPROVED** [review]

**Comment by andreeaflorescu** [comment]
I added the custom pipeline in Buildkite and retriggered the build. Any other PRs open against vm-fdt will fail the tests until this one is merged. If this becomes a problem I can also de-activate the custom build.

**Review by danielverkamp — APPROVED** [review]

**Comment by mkroening** [comment]
Thanks :)

Sorry for dismissing the reviews. I force-pushed the fix for the CI.

**Review by danielverkamp — APPROVED** [review]

File: .buildkite/custom-tests.json
Comment (andreeaflorescu) [inline .buildkite/custom-tests.json:5]: The rustup target add should actually go in the rust-vmm container. We can merge this as a temporary workaround, but we should eventually fix it. Do you want to send a PR with this change in rust-vmm-container as well? This is where we need to add the target: https://github.com/rust-vmm/rust-vmm-container/blob/main/build_container.sh#L40. If not, I can do that some time later.

Comment (mkroening) [reply .buildkite/custom-tests.json:5]: Makes sense. I opened https://github.com/rust-vmm/rust-vmm-container/pull/94.

Comment (andreeaflorescu) [reply .buildkite/custom-tests.json:5]: I am merging this PR, and once we publish a new version of rust-vmm-ci with the updated rust-vmm-container we can update this as well.

**Review by andreeaflorescu — APPROVED** [review]

File: .buildkite/custom-tests.json
Comment (mkroening) [reply .buildkite/custom-tests.json:5]: I opened https://github.com/rust-vmm/vm-fdt/pull/70 so we don't forget about this.

**Comment by mkroening** [comment]
Thanks a lot! When can we expect a release? :)

**Pull request merged**
